src/hotspot/share/classfile/javaClasses.hpp

```diff
@@ -532,10 +532,9 @@ class java_lang_VirtualThread : AllStatic {
     YIELDING     = 7,
     TERMINATED   = 99,
 
-    // can be suspended from scheduling when unmounted
-    SUSPENDED    = 1 << 8,
-    RUNNABLE_SUSPENDED = (RUNNABLE | SUSPENDED),
-    PARKED_SUSPENDED   = (PARKED | SUSPENDED)
+    // additional state bits
+    TIMED        = 1 << 8,   // timed parked
```

Comment:
Why handle TIMED this way instead of just making TIMED_PARKED/PARKING states?

Reply:
> Why handle TIMED this way instead of just making TIMED_PARKED/PARKING states?

Been there but it is 3 additional states and a lot simpler to just use the additional state bits, like we do with SUSPENDED. We might use another one for special casing sleeping, if the need arises.

Reply:
> Yes but TIMED and SUSPENDED are not alike. I assume SUSPENDED is a stand-alone bit because you can be in various states and suspended at the same time. But TIMED isn't like that.

The SUSPENDED bit can only be set when the virtual thread is unmounted, so limited to the RUNNABLE and PARKED states at this time.

The intention with TIMED was that it could be OR'ed with PARKING, PARKED or PINNED to quality that they are timed rather than untimed. It wouldn't make sense in other states.

Reply:
What is TIMED PINNING? 

To me TIMED_X are a specific set of states and there are not enough of them to consider TIMED to be a bit that can be applied to any state X.

Reply:
> What is TIMED PINNING?

LockSupport.parkNanos when the carrier can't be released due to monitor or native frame.

Reply:
Surely that is not a specified exported thread state though ?? Why would we care about PINNED (timed or otherwise) in the current context?

Reply:
> Surely that is not a specified exported thread state though ?? Why would we care about PINNED (timed or otherwise) in the current context?

There are internal states and then mappings to the thread states defined by the APIs (Thread::getState, JVMTI GetThreadState).  A virtual thread that parkNanos while pinned will park on its carrier, it doesn't unmount and release the carrier.  If some other threads polls the virtual thread state then will see Thread.State.TIMED_WAITING (or state bits that include JVMTI_THREAD_STATE_WAITING_WITH_TIMEOUT in the case of JVMTI). So no change to the API, no suggestion whatsoever of exposing "PINNED" in the API.